### ISSUE TITLE
SIGSEV when using range types with iterators and `move`

### ISSUE THREAD

**Issue by gmega** [nim-lang/Nim#23214]
### Description

The following snippet currently crashes in my system:

```nim
type 
  Result[T, E] = object
    case oResultPrivate*: bool
    of false:
      eResultPrivate*: E
    of true:
      vResultPrivate*: T

  NumType = Natural

  R = Result[NumType, ref CatchableError]

  ValueHolder = ref object of RootObj
    result*: R

proc transfer(holder: ValueHolder, value: sink R) = 
  holder.result = move(value)

var iter = iterator (holder: ValueHolder): R =
  var value = R(oResultPrivate: true, vResultPrivate: 1.Natural)
  transfer(holder, move(value))
  GC_fullCollect()

var holder = ValueHolder()

discard iter(holder)

echo $holder.result.vResultPrivate
```

### Nim Version

Nim Compiler Version 1.6.18 [Linux: amd64]
Compiled at 2023-12-15
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: a749a8b742bd0a4272c26a65517275db4720e58a
active boot switches: -d:release

This will also crash in every nim version from 2.0.2 down to 1.2.2, but won't crash in nim 1.2.0.

### Current Output

```text
Traceback (most recent call last)
/home/giuliano/Work/Status/chronos-tests/src/result_value_move.nim(18) result_value_move
/home/giuliano/Work/Status/chronos-tests/src/result_value_move.nim(14) :anonymous
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: '/home/giuliano/Work/Status/chronos-tests/build/result_value_move'
```

### Expected Output

```text
1
```

### Additional Information

Any of the following will fix the issue:

1. removing the call to `move` in transfer(holder, move(value)) (i.e. make it transfer(holder, value));
2. using `NumType = int` instead of `NumType = Natural`;
3. using nim 1.2.0;
4. using `--gc:arc` or `--gc:orc`.

In looking at the C code, I can see that the code that gets generated for the `move` within the iterator differs slightly when using `int` vs. `Natural`. For `int`, I get:
```nim
nimZeroMem((void *)(&T4_), sizeof(tyObject_Result__Ov9cK5Yc9bSGv7ieYfrM1duQ));
nimCopyMem((void *)(&T4_), (NIM_CONST void *)(&(*colonenvP_).value1), sizeof(T4_));
switch ((*colonenvP_).value1.oResultPrivate)
{
   case NIM_FALSE:
      unsureAsgnRef((void **)&(*colonenvP_).value1._oResultPrivate_1.eResultPrivate, NIM_NIL);
      break;
   case NIM_TRUE:
      (*colonenvP_).value1._oResultPrivate_2.vResultPrivate = 0;
      break;
}
(*colonenvP_).value1.oResultPrivate = 0;
```
for `Natural`, instead, I get:
```nim
nimZeroMem((void *)(&T4_), sizeof(tyObject_Result__049by0G76bdqNtgv2jXPl9aA));
nimCopyMem((void *)(&T4_), (NIM_CONST void *)(&(*colonenvP_).value1), sizeof(T4_));
switch ((*colonenvP_).value1.oResultPrivate)
{
   case NIM_FALSE:
      unsureAsgnRef((void **)&(*colonenvP_).value1._oResultPrivate_1.eResultPrivate, NIM_NIL);
      break;
   case NIM_TRUE:
      break;
}
(*colonenvP_).value1.oResultPrivate = 0;
```
which means the range type is missing the:
```nim
(*colonenvP_).value1._oResultPrivate_2.vResultPrivate = 0;
```
Putting that line in the C code manually and compiling fixes the crash.

Bisecting led me to this commit: 7800fa394f26e9a5a6cb30964ba005512199be58, which changes the part of the code generator which produces the "reset" code for objects, which seems to match well with the role of the missing line. Why this leads to a SIGSEV is beyond me, however. :-) Probably related to #22462 in some way.

**Comment by gmega** [comment]
I think I may have found a way to fix it. Will file a PR tmw.

**Comment by ringabout** [comment]
fixed by #23215
